src/EditorFeatures/Core/CommandHandlers/AbstractGoToCommandHandler`2.cs

```diff
@@ -0,0 +1,259 @@
+        // and failure ourselves.
+        try
+        {
+            // Should only be called on the UI thread.
+            Contract.ThrowIfFalse(_threadingContext.HasMainThread);
+
+            // Make an tracking token so that integration tests can wait until we're complete.
+            using var token = _listener.BeginAsyncOperation($"{this.GetType().Name}.{nameof(ExecuteCommandAsync)}");
+
+            // Only start running once the previous command has finished.  That way we don't have results from both
+            // potentially interleaving with each other.  Note: this should ideally always be fast as long as the prior
+            // task respects cancellation.
+            //
+            // Note: we just need to make sure we run after that prior command finishes.  We do not want to propagate
+            // any failures from it.  Technically this should not be possible as it should be inside this same
+            // try/catch. however this code wants to be very resilient to any prior mistakes infecting later operations.
+            await _inProgressCommand.NoThrowAwaitable(captureContext: false);
+            await this.ExecuteCommandWorkerAsync(document, position, cancellationTokenSource).ConfigureAwait(false);
+        }
+        catch (OperationCanceledException)
+        {
+        }
+        catch (Exception ex) when (FatalError.ReportAndCatch(ex))
+        {
+        }
+    }
+
+    private async Task ExecuteCommandWorkerAsync(
+        Document document,
+        int position,
+        CancellationTokenSource cancellationTokenSource)
+    {
+        // Switch to the BG immediately so we can keep as much work off the UI thread.
+        await TaskScheduler.Default;
+
+        // We kick off the work to find the impl/base in the background.  If we get the results for it within 1.5
+        // seconds, we then either navigate directly to it (in the case of one result), or we show all the results in
+        // the presenter (in the case of multiple).
+        //
+        // However, if the results don't come back in 1.5 seconds, we just pop open the presenter and continue the
+        // search there.  That way the user is not blocked and can go do other work if they want.
+
+        // We create our own context object, simply to capture all the definitions reported by the individual
+        // TLanguageService.  Once we get the results back we'll then decide what to do with them.  If we get only a
+        // single result back, then we'll just go directly to it.  Otherwise, we'll present the results in the
+        // IStreamingFindUsagesPresenter.
+        var findContext = new BufferedFindUsagesContext();
+
+        var cancellationToken = cancellationTokenSource.Token;
+        var delayTask = Task.Delay(TaggerDelay.OnIdle.ComputeTimeDelay(), cancellationToken);
+        var findTask = Task.Run(() => FindResultsAsync(document, position, findContext, cancellationToken), cancellationToken);
```

Comment:
For my curiosity: Is the `Task.Run` needed here, if we're already on a background thread due to the TaskScheduler.Default?

Reply:
It likely isn't.  i can remove in a followup :)